File: linked_file_allocation.py
#!/usr/bin/env python3
"""Linked File Allocation - Each block points to the next block"""

def linked_allocation(files, disk_size):
    """
    files: list of (filename, start_block, num_blocks)
    disk_size: total number of blocks
    """
    disk = [-1] * disk_size  # -1 means free, otherwise stores next block number
    allocation = {}
    
    for filename, start, num_blocks in files:
        if start >= disk_size or start < 0:
            print(f"Error: Invalid start block {start} for {filename}")
            continue
        
        # check if start block is free
        if disk[start] != -1:
            print(f"Error: Block {start} already allocated for {filename}")
            continue
        
        # allocate blocks
        blocks = [start]
        current = start
        
        for i in range(1, num_blocks):
            # find next free block
            next_block = None
            for j in range(disk_size):
                if disk[j] == -1 and j not in blocks:
                    next_block = j
                    break
            
            if next_block is None:
                print(f"Error: Not enough free blocks for {filename}")
                break
            
            disk[current] = next_block
            blocks.append(next_block)
            current = next_block
        
        disk[current] = -2  # -2 means end of file
        allocation[filename] = blocks
    
    return allocation, disk


def print_allocation(allocation):
    print("\nLinked File Allocation:")
    print(f"{'File':<15} {'Blocks'}")
    print("-" * 40)
    for filename, blocks in allocation.items():
        blocks_str = " -> ".join(map(str, blocks))
        print(f"{filename:<15} {blocks_str}")


if __name__ == '__main__':
    files = [
        ("file1.txt", 0, 3),
        ("file2.txt", 5, 4),
        ("file3.txt", 10, 2),
    ]
    disk_size = 20
    
    allocation, disk = linked_allocation(files, disk_size)
    print_allocation(allocation)
    
    print(f"\nDisk Status (showing links):")
    for i, link in enumerate(disk):
        if link == -1:
            status = "Free"
        elif link == -2:
            status = "End"
        else:
            status = f"-> {link}"
        if link != -1:
            print(f"Block {i}: {status}")

File: linked_file_allocation.js
#!/usr/bin/env node
/**
 * Linked File Allocation - Each block points to the next block
 */

function linkedAllocation(files, diskSize) {
    const disk = new Array(diskSize).fill(-1); // -1 means free
    const allocation = {};
    
    for (const file of files) {
        const { filename, start, numBlocks } = file;
        
        if (start >= diskSize || start < 0) {
            console.log(`Error: Invalid start block ${start} for ${filename}`);
            continue;
        }
        
        // check if start block is free
        if (disk[start] !== -1) {
            console.log(`Error: Block ${start} already allocated for ${filename}`);
            continue;
        }
        
        // allocate blocks
        const blocks = [start];
        let current = start;
        
        for (let i = 1; i < numBlocks; i++) {
            // find next free block
            let nextBlock = null;
            for (let j = 0; j < diskSize; j++) {
                if (disk[j] === -1 && !blocks.includes(j)) {
                    nextBlock = j;
                    break;
                }
            }
            
            if (nextBlock === null) {
                console.log(`Error: Not enough free blocks for ${filename}`);
                break;
            }
            
            disk[current] = nextBlock;
            blocks.push(nextBlock);
            current = nextBlock;
        }
        
        disk[current] = -2; // -2 means end of file
        allocation[filename] = blocks;
    }
    
    return { allocation, disk };
}

function printAllocation(allocation) {
    console.log("\nLinked File Allocation:");
    console.log("File".padEnd(15), "Blocks");
    console.log("-".repeat(40));
    
    for (const [filename, blocks] of Object.entries(allocation)) {
        const blocksStr = blocks.join(" -> ");
        console.log(filename.padEnd(15), blocksStr);
    }
}

// Main execution
if (require.main === module) {
    const files = [
        { filename: "file1.txt", start: 0, numBlocks: 3 },
        { filename: "file2.txt", start: 5, numBlocks: 4 },
        { filename: "file3.txt", start: 10, numBlocks: 2 }
    ];
    const diskSize = 20;
    
    const { allocation, disk } = linkedAllocation(files, diskSize);
    printAllocation(allocation);
    
    console.log("\nDisk Status (showing links):");
    for (let i = 0; i < disk.length; i++) {
        if (disk[i] !== -1) {
            const status = disk[i] === -2 ? "End" : `-> ${disk[i]}`;
            console.log(`Block ${i}: ${status}`);
        }
    }
}

module.exports = { linkedAllocation, printAllocation };
